<a href="https://colab.research.google.com/github/bhushanmandava/Prediction-of-Alzheimer-s-Disease/blob/main/Alzheimer's_Disease_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Goal is to predecit that the patient have Alzhemier or not

## DATA PreProcessing

### importing the required libraries

In [ ]:
import pandas as pd
import numpy as np

### importing the dataset

In [ ]:
dataset =pd.read_csv("alzheimers_disease_data.csv")

### Checking the data set info

In [ ]:
dataset.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2149 entries, 0 to 2148
Data columns (total 35 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   PatientID                  2149 non-null   int64  
 1   Age                        2149 non-null   int64  
 2   Gender                     2149 non-null   int64  
 3   Ethnicity                  2149 non-null   int64  
 4   EducationLevel             2149 non-null   int64  
 5   BMI                        2149 non-null   float64
 6   Smoking                    2149 non-null   int64  
 7   AlcoholConsumption         2149 non-null   float64
 8   PhysicalActivity           2149 non-null   float64
 9   DietQuality                2149 non-null   float64
 10  SleepQuality               2149 non-null   float64
 11  FamilyHistoryAlzheimers    2149 non-null   int64  
 12  CardiovascularDisease      2149 non-null   int64  
 13  Diabetes                   2149 non-null   int64

#### **Insights**
Now with the data info we got we can say we have 35 total coloumns and 2149 entries and also we donot have any Missing Data in any of the coloumns

In [ ]:
dataset.head()

PatientID  Age  Gender  Ethnicity  EducationLevel        BMI  Smoking  \
0       4751   73       0          0               2  22.927749        0   
1       4752   89       0          0               0  26.827681        0   
2       4753   73       0          3               1  17.795882        0   
3       4754   74       1          0               1  33.800817        1   
4       4755   89       0          0               0  20.716974        0   

   AlcoholConsumption  PhysicalActivity  DietQuality  ...  MemoryComplaints  \
0           13.297218          6.327112     1.347214  ...                 0   
1            4.542524          7.619885     0.518767  ...                 0   
2           19.555085          7.844988     1.826335  ...                 0   
3           12.209266          8.428001     7.435604  ...                 0   
4           18.454356          6.310461     0.795498  ...                 0   

   BehavioralProblems       ADL  Confusion  Disorientation  \
0                   0  1.725883          0               0   
1                   0  2.592424          0               0   
2                   0  7.119548          0               1   
3                   1  6.481226          0               0   
4                   0  0.014691          0               0   

   PersonalityChanges  DifficultyCompletingTasks  Forgetfulness  Diagnosis  \
0                   0                          1              0          0   
1                   0                          0              1          0   
2                   0                          1              0          0   
3                   0                          0              0          0   
4                   1                          1              0          0   

   DoctorInCharge  
0       XXXConfid  
1       XXXConfid  
2       XXXConfid  
3       XXXConfid  
4       XXXConfid  

[5 rows x 35 columns]

#### Insights
Now when we are see the data the column DoctorInCharge is confidential and also it is irrelevant to our model so we can drop it

### droping DoctorInCharge

In [ ]:
dataset =dataset.drop("DoctorInCharge",axis=1)


In [ ]:
dataset.head()

PatientID  Age  Gender  Ethnicity  EducationLevel        BMI  Smoking  \
0       4751   73       0          0               2  22.927749        0   
1       4752   89       0          0               0  26.827681        0   
2       4753   73       0          3               1  17.795882        0   
3       4754   74       1          0               1  33.800817        1   
4       4755   89       0          0               0  20.716974        0   

   AlcoholConsumption  PhysicalActivity  DietQuality  ...  \
0           13.297218          6.327112     1.347214  ...   
1            4.542524          7.619885     0.518767  ...   
2           19.555085          7.844988     1.826335  ...   
3           12.209266          8.428001     7.435604  ...   
4           18.454356          6.310461     0.795498  ...   

   FunctionalAssessment  MemoryComplaints  BehavioralProblems       ADL  \
0              6.518877                 0                   0  1.725883   
1              7.118696                 0                   0  2.592424   
2              5.895077                 0                   0  7.119548   
3              8.965106                 0                   1  6.481226   
4              6.045039                 0                   0  0.014691   

   Confusion  Disorientation  PersonalityChanges  DifficultyCompletingTasks  \
0          0               0                   0                          1   
1          0               0                   0                          0   
2          0               1                   0                          1   
3          0               0                   0                          0   
4          0               0                   1                          1   

   Forgetfulness  Diagnosis  
0              0          0  
1              1          0  
2              0          0  
3              0          0  
4              0          0  

[5 rows x 34 columns]

In [ ]:
x=dataset.iloc[:,1:-1].values
y=dataset.iloc[:,-1].values


In [ ]:
print(x)

[[73.  0.  0. ...  0.  1.  0.]
 [89.  0.  0. ...  0.  0.  1.]
 [73.  0.  3. ...  0.  1.  0.]
 ...
 [77.  0.  0. ...  0.  0.  0.]
 [78.  1.  3. ...  0.  0.  1.]
 [72.  0.  0. ...  0.  0.  1.]]


### Getting the matrix of Features

#### before getting our best feature we gonna scale the features

In [ ]:
# from sklearn.preprocessing import StandardScaler
# sc=StandardScaler()
# x=sc.fit_transform(x)

In [ ]:
# from sklearn.decomposition import KernelPCA
# kpca = KernelPCA(n_components=15,kernel="rbf")
# x=kpca.fit_transform(x)

In [ ]:
from sklearn.feature_selection import SelectKBest, f_classif
sb =SelectKBest(score_func=f_classif,k=10)
x=sb.fit_transform(dataset.drop(["Diagnosis"],axis=1),y)

In [ ]:
sb.get_feature_names_out()

array(['PatientID', 'EducationLevel', 'SleepQuality', 'Hypertension',
       'CholesterolHDL', 'MMSE', 'FunctionalAssessment',
       'MemoryComplaints', 'BehavioralProblems', 'ADL'], dtype=object)

#### we got the 10 best features for the x now we can go into model training


In [ ]:
x.shape[1]
x

array([[4.75100000e+03, 2.00000000e+00, 9.02567867e+00, ...,
        0.00000000e+00, 0.00000000e+00, 1.72588346e+00],
       [4.75200000e+03, 0.00000000e+00, 7.15129274e+00, ...,
        0.00000000e+00, 0.00000000e+00, 2.59242413e+00],
       [4.75300000e+03, 1.00000000e+00, 9.67357416e+00, ...,
        0.00000000e+00, 0.00000000e+00, 7.11954774e+00],
       ...,
       [6.89700000e+03, 1.00000000e+00, 5.76946355e+00, ...,
        0.00000000e+00, 0.00000000e+00, 5.03633399e+00],
       [6.89800000e+03, 1.00000000e+00, 8.32287396e+00, ...,
        0.00000000e+00, 0.00000000e+00, 3.78539871e+00],
       [6.89900000e+03, 2.00000000e+00, 9.87871052e+00, ...,
        0.00000000e+00, 1.00000000e+00, 8.32756301e+00]])

### spliting our data

In [ ]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=0)

In [ ]:
from sklearn.ensemble import GradientBoostingClassifier
model = GradientBoostingClassifier()
model.fit(x_train,y_train)

GradientBoostingClassifier()

In [ ]:
y_pred=model.predict(x_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1),y_test.reshape(len(y_test),1)),1))

[[0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [0 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [1 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 1]
 [1 1]
 [1 1]
 [0 0]
 [0 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [1 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]

### Getting our accuracy

In [ ]:
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)
print("Accuracy: {:.2f} %".format(accuracy*100))
print()
print("Confusion Matrix:")
print(conf_matrix)
print()
print("Classification Report:")
print(class_report)

Accuracy: 96.05 %

Confusion Matrix:
[[258   9]
 [  8 155]]

Classification Report:
              precision    recall  f1-score   support

           0       0.97      0.97      0.97       267
           1       0.95      0.95      0.95       163

    accuracy                           0.96       430
   macro avg       0.96      0.96      0.96       430
weighted avg       0.96      0.96      0.96       430



### hyper parameters tuning

In [ ]:
from sklearn.model_selection import cross_val_score, GridSearchCV, StratifiedKFold

param_grid = {
    'n_estimators': [50, 100, 200],
    'learning_rate': [0.01, 0.1, 0.2],
    'max_depth': [3, 5, 7]
}

In [ ]:
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=10, scoring='accuracy')
grid_search.fit(x_train, y_train)

GridSearchCV(cv=10, estimator=GradientBoostingClassifier(),
             param_grid={'learning_rate': [0.01, 0.1, 0.2],
                         'max_depth': [3, 5, 7],
                         'n_estimators': [50, 100, 200]},
             scoring='accuracy')

In [ ]:
best_model =grid_search.best_estimator_
best_params=grid_search.best_params_

In [ ]:
best_params

{'learning_rate': 0.2, 'max_depth': 3, 'n_estimators': 200}

In [ ]:
best_model.fit(x_train,y_train)

GradientBoostingClassifier(learning_rate=0.2, n_estimators=200)

In [ ]:
best_prediction = best_model.predict(x_test)

In [ ]:
accuracy = accuracy_score(y_test, best_prediction)
conf_matrix = confusion_matrix(y_test, best_prediction)
class_report = classification_report(y_test, best_prediction)
print("Accuracy: {:.2f} %".format(accuracy*100))
print()
print("Confusion Matrix:")
print(conf_matrix)
print()
print("Classification Report:")
print(class_report)

Accuracy: 96.51 %

Confusion Matrix:
[[260   7]
 [  8 155]]

Classification Report:
              precision    recall  f1-score   support

           0       0.97      0.97      0.97       267
           1       0.96      0.95      0.95       163

    accuracy                           0.97       430
   macro avg       0.96      0.96      0.96       430
weighted avg       0.97      0.97      0.97       430



### Getting Cross validation scores

In [ ]:
cv_scores = cross_val_score(best_model, x, y, cv=5)

In [ ]:
print("K-Fold Cross Validation Accuracy: {:.2f} %".format(cv_scores.mean()*100))

K-Fold Cross Validation Accuracy: 95.43 %


In [ ]:
print("K-Fold Cross Validation Standard Deviation: {:.2f} %".format(cv_scores.std()*100))

K-Fold Cross Validation Standard Deviation: 7.17 %


In [ ]:
skf = StratifiedKFold(n_splits=10, shuffle=True)
skf_cv_scores = cross_val_score(best_model, x, y, cv=skf)

In [ ]:
print("Stratified K-Fold Cross Validation Accuracy: {:.2f} %".format(skf_cv_scores.mean()*100))

Stratified K-Fold Cross Validation Accuracy: 95.35 %


In [ ]:
print("Stratified K-Fold Cross Validation Standard Deviation: {:.2f} %".format(skf_cv_scores.std()*100))

Stratified K-Fold Cross Validation Standard Deviation: 0.97 %
